lib/tree_construction/location_info_mixin.js

```diff
@@ -10,7 +10,19 @@ var $ = HTML.TAG_NAMES;
 
 
 function setEndLocation(element, endTagToken) {
-    if (element.__location)
+    if(!element.__startTagLocation && element.__location) {
+        element.__startTagLocation = {
+            start: element.__location.start,
+            end: element.__location.end
+        };
+    }
+    if(endTagToken.location && endTagToken.type === Tokenizer.END_TAG_TOKEN && element.tagName === endTagToken.tagName) {
```

Comment:
Yes, both of them are required.
First one handles cases like
<code>&lt;p&gt; &lt;p&gt; &lt;/p&gt;</code>
and ensures no __endTagLocation for the first &lt;p&gt; (because it doesn't have one)

The second one handles cases like
<code>&lt;table&gt;</code>
<code>&nbsp;&nbsp;&lt;tr&gt;</code>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&lt;td&gt;</code>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;p&gt;</code>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&lt;/td&gt;</code>
<code>&nbsp;&nbsp;&lt;/tr&gt;</code>
<code>&lt;/table&gt;</code>
and ensures no __endTagLocation for &lt;p&gt;